presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,793 @@
+            }
+
+            return INSTANCE_SIZE + sizeOf(positions) + sizeOf(rowSizes) + buffersSizeInBytes;
+        }
+
+        private void initializeBlockEncodingBuffers(DecodedBlockNode[] decodedBlocks)
+        {
+            // Create buffers has to be done after seeing the first page.
+            if (blockEncodingBuffers == null) {
+                int initialElementCount = min(capacity / channelCount / DEFAULT_ELEMENT_SIZE_IN_BYTES, INITIAL_POSITION_COUNT);
+
+                blockEncodingBuffers = new BlockEncodingBuffers[channelCount];
+                for (int i = 0; i < channelCount; i++) {
+                    blockEncodingBuffers[i] = createBlockEncodingBuffers(decodedBlocks[i], initialElementCount);
+                }
+            }
+        }
+
+        private void calculateRowSizes(int fixedWidthRowSize, List<Integer> variableWidthChannels)
+        {
+            if (variableWidthChannels.isEmpty()) {
+                return;
+            }
+
+            ensureRowSizesCapacityAndInitialize();
+
+            for (int i : variableWidthChannels) {
+                blockEncodingBuffers[i].accumulateRowSizes(rowSizes);
+            }
+
+            for (int i = 0; i < positionCount; i++) {
+                rowSizes[i] += fixedWidthRowSize;
+            }
+        }
+
+        private int calculateNextBatchSize(int fixedWidthRowSize, List<Integer> variableWidthChannels, int startPosition)
+        {
+            int bytesRemaining = capacity - toIntExact(getSerializedBuffersSizeInBytes());
+
+            if (variableWidthChannels.isEmpty()) {
+                int positionsFit = max(bytesRemaining / fixedWidthRowSize, 1);
+                bufferFull = (positionsFit <= positionCount - startPosition);
+                return min(positionsFit, positionCount - startPosition);
+            }
+
+            verify(rowSizes != null);
+            for (int i = startPosition; i < positionCount; i++) {
+                bytesRemaining -= rowSizes[i];
+
+                if (bytesRemaining <= 0) {
+                    bufferFull = true;
+                    return max(i - startPosition, 1);
+                }
+            }
+
+            return positionCount - startPosition;
+        }
+
+        private void flush(OutputBuffer outputBuffer)
+        {
+            if (bufferedRowCount == 0) {
+                return;
+            }
+
+            SliceOutput output = new DynamicSliceOutput(toIntExact(getSerializedBuffersSizeInBytes()));
+            output.writeInt(channelCount);
+
+            for (int i = 0; i < channelCount; i++) {
+                blockEncodingBuffers[i].serializeTo(output);
+            }
+
+            SerializedPage serializedPage = serde.serialize(output.slice(), bufferedRowCount);
+            outputBuffer.enqueue(lifespan, partition, ImmutableList.of(serializedPage));
+            pagesAdded.incrementAndGet();
+            rowsAdded.addAndGet(bufferedRowCount);
+
+            if (blockEncodingBuffers != null) {
```

Comment:
this logic is inconsistent with the for loop above; change it to

```
            for (int i = 0; i < channelCount; i++) {
                blockEncodingBuffers[i].resetBuffers();
            }
```

Reply:
Why this comment is marked as resolved? Since on line 477 `blockEncodingBuffers` is not checked whether it's null or not, so I assume here we don't need the check as well?

Also, does it make sense to have one single `serializeAndReset` (a.k.a `flush`) API ? As it looks to me `serializeTo` call is always coupled with `resetBuffers`. -- We can do this refactor after this PR is merged, of course.